pkg/status/singletonstatus.go

```diff
@@ -0,0 +1,167 @@
+	logger := klog.FromContext(ctx)
+	logger.Info("Initializing the desired state for singleton statuses")
+
+	allBdgs, err := c.wdsKsClient.ControlV1alpha1().Bindings().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, bdg := range allBdgs.Items {
+		logger.V(2).Info("Got Binding", "name", bdg.Name)
+		list := bdg.Spec.Workload.NamespaceScope // Should also do the same thing for cluster scoped
+		for _, item := range list {
+			nsWObjRef := item.NamespaceScopeDownsyncObject
+			lister, found := c.listers.Get(metav1GVR2schemaGVR(nsWObjRef.GroupVersionResource))
+			if !found {
+				return fmt.Errorf("could not get lister for gvr: %s", nsWObjRef.GroupVersionResource)
+			}
+			nsWObj, err := lister.ByNamespace(nsWObjRef.Namespace).Get(nsWObjRef.Name)
+			if err != nil {
+				return err
+			}
+			labels := nsWObj.(metav1.Object).GetLabels()
+			if v, ok := labels[util.BindingPolicyLabelSingletonStatusKey]; ok {
+				logger.V(2).Info("Found namespace scoped singleton workload object", "name", nsWObjRef.Name)
+				wObjIdentifier := util.ObjectIdentifier{
+					GVK:        schema.GroupVersionKind{Group: nsWObjRef.Group, Version: nsWObjRef.Version, Kind: nsWObj.GetObjectKind().GroupVersionKind().Kind},
+					Resource:   nsWObjRef.Resource,
+					ObjectName: cache.NewObjectName(nsWObjRef.Namespace, nsWObjRef.Name),
+				}
+				if v == util.BindingPolicyLabelSingletonStatusValueSet && len(bdg.Spec.Destinations) > 0 {
+					logger.V(2).Info("Singleton workload object should has status synced", "name", nsWObjRef.Name)
+					theSingletonState.Set(wObjIdentifier)
+				} else {
+					logger.V(2).Info("Singleton workload object should not have status synced", "name", nsWObjRef.Name)
+					theSingletonState.Unset(wObjIdentifier)
+				}
+				// reconcile should be out of the for loop over all bindings
+				// because the desired state for a single workload object must be determined over ALL bindings
+				if reconcile {
+					c.reconcileSingletonByWObj(ctx, wObjIdentifier)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func metav1GVR2schemaGVR(in metav1.GroupVersionResource) (out schema.GroupVersionResource) {
+	out = schema.GroupVersionResource{
+		Group:    in.Group,
+		Version:  in.Version,
+		Resource: in.Resource,
+	}
+	return
+}
+
+func (c *Controller) reconcileSingletonByWObj(ctx context.Context, wObjID util.ObjectIdentifier) error {
+	logger := klog.FromContext(ctx)
+
+	if !theSingletonState.Get(wObjID) {
+		emptyStatus := make(map[string]interface{})
+		return updateObjectStatus(ctx, &wObjID, emptyStatus, c.listers, c.wdsDynClient)
+	}
+
+	wsNameSuffix := strings.Join([]string{
+		wObjID.ObjectName.Namespace,
+		wObjID.ObjectName.Name,
+	}, "-")
+	// A (probably critical) question: How do I know which WEC is selected?
+	list, _ := c.workStatusLister.ByNamespace("cluster1").List(labels.Everything())
+	for _, obj := range list {
+		wsName := obj.(metav1.Object).GetName()
+		if !strings.HasSuffix(wsName, wsNameSuffix) {
```

Comment:
Where is the construction of WorkStatus object names documented?
It looks to me like the current practice has the problems of possible name collisions and length overflows.
It would be better if this client code got its job done without depending on how names are constructed.
Why not look at the source object reference in the spec of the WorkStatus?